mubench.pipeline/data/detector.py

```diff
@@ -1,33 +1,56 @@
+from logging import Logger
 from os.path import join, exists
-from typing import Optional, List
+from typing import Dict, Optional, List
 
 from data.finding import Finding, SpecializedFinding
-
+from data.project_version import ProjectVersion
+from data.runner_interface import RunnerInterface
+from utils.io import read_yaml
 
 class Detector:
     BASE_URL = "http://www.st.informatik.tu-darmstadt.de/artifacts/mubench/detectors"
-    CLI_VERSION = "v20170406"
+    RELEASES_FILE = "releases.yml"
 
-    def __init__(self, detectors_path: str, detector_id: str, java_options: List[str]):
+    def __init__(self, detectors_path: str, detector_id: str, java_options: List[str], requested_release: Optional[str] = None):
         self.id = detector_id
         self.base_name = detector_id.split("_", 1)[0]
-        self.java_options = java_options
-
         self.path = join(detectors_path, self.id)
+
+        releases_index_path = join(self.path, Detector.RELEASES_FILE)
+        release = self._get_release(releases_index_path, requested_release)
+        release_tag = release.get("tag", "latest")
+
+        if "cli_version" in release:
+            cli_version = release["cli_version"]
+        else:
+            raise ValueError("Missing CLI version for {}".format(detector_id))
+
+        if "md5" in release:
+            self.md5 = release["md5"]
+        else:
+            raise ValueError("Missing MD5 for {}".format(detector_id))
+
         self.jar_path = join(self.path, self.base_name + ".jar")
-        self.jar_url = "{}/{}/{}.jar".format(Detector.BASE_URL, Detector.CLI_VERSION, self.base_name)
-        self.md5_path = join(self.path, self.base_name + ".md5")
+        self.jar_url = "{}/{}/{}/{}.jar".format(Detector.BASE_URL, release_tag, cli_version, self.base_name)
+
+        self.runner_interface = RunnerInterface.get(cli_version, self.jar_path, java_options)
+
+    def _get_release(self, releases_index_file_path: str, requested_release: Optional[str]) -> Dict[str, str]:
+        if exists(releases_index_file_path):
+            releases = read_yaml(releases_index_file_path)
+            if requested_release:
+                releases = [r for r in releases if r.get("tag", "latest") == requested_release]
 
-    @property
-    def md5(self) -> Optional[str]:
-        md5_file = self.md5_path
-        md5 = None
+            if releases:
+                return releases[0]
+            else:
+                raise ValueError("No matching {} release for {}".format(self.id, requested_release))
 
-        if exists(md5_file):
-            with open(md5_file) as file:
-                md5 = file.read().strip()
+        return dict()
```

Comment:
Shouldn't this case also raise an exception, because detectors need to specify releases now?